http/events_test.go

```diff
@@ -201,3 +208,76 @@ func TestEventsSubscribeAuth(t *testing.T) {
 		t.Errorf("Expected 403 but got %+v", resp)
 	}
 }
+
+func TestCanForwardEventConnections(t *testing.T) {
+	// Run again with in-memory network
+	inmemCluster, err := cluster.NewInmemLayerCluster("inmem-cluster", 3, hclog.New(&hclog.LoggerOptions{
+		Mutex: &sync.Mutex{},
+		Level: hclog.Trace,
+		Name:  "inmem-cluster",
+	}))
+	if err != nil {
+		t.Fatal(err)
+	}
+	testCluster := vault.NewTestCluster(t, &vault.CoreConfig{
+		Experiments: []string{experiments.VaultExperimentEventsAlpha1},
+		AuditBackends: map[string]audit.Factory{
+			"nop": corehelpers.NoopAuditFactory(nil),
+		},
+	}, &vault.TestClusterOptions{
+		ClusterLayers: inmemCluster,
+	})
+	cores := testCluster.Cores
+	testCluster.Start()
+	defer testCluster.Cleanup()
+
+	rootToken := testCluster.RootToken
+
+	// Wait for core to become active
+	vault.TestWaitActiveForwardingReady(t, cores[0].Core)
+
+	// Test forwarding a request. Since we're going directly from core to core
+	// with no fallback we know that if it worked, request handling is working
+	c := cores[1]
+	standby, err := c.Standby()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !standby {
+		t.Fatal("expected core to be standby")
+	}
+
+	// We need to call Leader as that refreshes the connection info
+	isLeader, _, _, err := c.Leader()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if isLeader {
+		t.Fatal("core should not be leader")
+	}
+	corehelpers.RetryUntil(t, 5*time.Second, func() error {
+		state := c.ActiveNodeReplicationState()
+		if state == 0 {
+			return fmt.Errorf("heartbeats have not yet returned a valid active node replication state: %d", state)
+		}
+		return nil
+	})
+
+	req, err := http.NewRequest("GET", "https://pushit.real.good:9281/v1/sys/events/subscribe/xyz?json=true", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	req = req.WithContext(namespace.RootContext(req.Context()))
+	req.Header.Add(consts.AuthHeaderName, rootToken)
+
+	resp := httptest.NewRecorder()
+	forwardRequest(cores[1].Core, resp, req)
```

Comment:
This seems like a good test for ensuring the correct behaviour given a websocket request that needs forwarding. Is it practical to have another slightly more zoomed out test that will exercise the logic on whether a websocket request needs forwarding or not? For example, loop over 3 cores sending a request to each and see that 1 gets served directly, and the other 2 get redirected.

Reply:
I believe I can do that.